### DISCUSSION TITLE
DeDRM Decryption of EPUB in Calibre failed - Exception when getting Adobe Key

### DISCUSSION THREAD

**Discussion by DRWolff1954** [apprenticeharper/DeDRM_tools#1798] [Q&A]
Like so many others, I ran into problems with the Calibre update to version 5. Unlike many others, I have not found any solution that works.
I tried downgrading to Calibre 4.23 with DeDRM 6.8.1. Didn't work.
I tried going to even earlier versions of Calibre/DeDRM. Didn't work.
I tried Calibre Portable 4.2.3 with DeDRM 6.8.1. Didn't work.
Most recently, I made sure that I had the most recent versions of Adobe Digital Edition (V4.5), Calibre-64bit (V5.2.4) and DeDRM (V7.2.1). 
I then downloaded a brand-new EPUB book (well, an ascm link) that I had purchased from Kobo. It opened in ADE, and I was able to read it with no problem.
I closed ADE and opened Calibre. I tried to add the new book. Didn't work. Still DRM-locked.
Puzzling, and irritating. 
I hope someone can help solve this.

Below are copied debug info from two sources
1) DETAILS OF IMPORT FAILURE GENERATED USING COMMAND WINDOW AND "calibre-debug -g" (without the quotes)
2) DETAILS OF IMPORT FAILURE GENERATED IN CALIBRE.

Thanks in advance
DRWolff1954

**_1) DETAILS OF DECRYPT FAILURE GENERATED USING COMMAND WINDOW AND "calibre-debug -g" (without the quotes)_**

C:\WINDOWS\system32>calibre-debug -g
calibre 5.24 [64bit]  embedded-python: True is64bit: True
Windows-10-10.0.19041 Windows ('64bit', 'WindowsPE')
('Windows', '10', '10.0.19041')
Python 3.8.5
Windows: ('10', '10.0.19041', '', 'Multiprocessor Free')
Interface language: None
Successfully initialized third party plugins: DeDRM (7, 2, 1)
Turning on automatic hidpi scaling
devicePixelRatio: 1.0
logicalDpi: 96.0 x 96.0
physicalDpi: 94.14671814671814 x 94.07407407407408
Using calibre Qt style: True
[0.00] Starting up...
[0.00] Showing splash screen...
[0.41] splash screen shown
[0.41] Initializing db...
[0.44] db initialized
[0.44] Constructing main UI...
[1.02] main UI initialized...
[1.02] Hiding splash screen
Starting QuickView
[1.28] splash screen hidden
[1.28] Started up in 1.28 seconds with 650 books
stdout+stderr from file dialog helper: [b'', b'']
piped data from file dialog helper: [b'k%\x87u\x90\xfb\x06\xf9\xaa\x98\x96\x90\xa5K\xe5/ \xfd\x9f\xa8ic\xc0\xacc(M;\x0e\x08"\x84', b'D:\\Users\\Denis\\DenisDocs\\My Digital Editions\\Velocity Weapon.epub']
DeDRM v7.2.1: Trying to decrypt Velocity Weapon.epub
DeDRM v7.2.1: Verifying zip archive integrity
DeDRM v7.2.1: Velocity Weapon.epub is a secure Adobe Adept ePub
DeDRM v7.2.1: Trying Encryption key default_key
Could not decrypt 5c_g6c66.epub because of an exception:
Traceback (most recent call last):
  File "calibre_plugins.dedrm.ineptepub", line 418, in decryptBook
TypeError: object of type 'int' has no len()

DeDRM v7.2.1: Failed to decrypt with key default_key after 0.1 seconds
DeDRM v7.2.1: Looking for new default Adobe Digital Editions Keys after 0.1 seconds
DeDRM v7.2.1: Exception when getting default Adobe Key after 0.1 seconds
Traceback (most recent call last):
  File "calibre_plugins.dedrm.__init__", line 358, in ePubDecrypt
  File "calibre_plugins.dedrm.adobekey", line 362, in adeptkeys
  File "calibre_plugins.dedrm.adobekey", line 342, in CryptUnprotectData
calibre_plugins.dedrm.adobekey.ADEPTError: Failed to decrypt user key key (sic)
DeDRM v7.2.1: Ultimately failed to decrypt after 0.1 seconds. Read the FAQs at Harper's repository: https://github.com/apprenticeharper/DeDRM_tools/blob/master/FAQs.md
Added Velocity Weapon to db in: 1.7
Added 1 books in 2.2 seconds
Running file type plugin DeDRM failed with traceback:
Traceback (most recent call last):
  File "calibre\customize\ui.py", line 176, in _run_filetype_plugins
  File "calibre_plugins.dedrm.__init__", line 644, in run
  File "calibre_plugins.dedrm.__init__", line 415, in ePubDecrypt
calibre_plugins.dedrm.DeDRMError: DeDRM v7.2.1: Ultimately failed to decrypt after 0.1 seconds. Read the FAQs at Harper's repository: https://github.com/apprenticeharper/DeDRM_tools/blob/master/FAQs.md
Exception ignored in: <function NativeFunction.__del__ at 0x000001D4F96CB430>
Traceback (most recent call last):
  File "calibre_plugins.dedrm.adobekey", line 268, in __del__
TypeError: 'NoneType' object is not callable
Exception ignored in: <function NativeFunction.__del__ at 0x000001D4F96CB430>
Traceback (most recent call last):
  File "calibre_plugins.dedrm.adobekey", line 268, in __del__
TypeError: 'NoneType' object is not callable
stdout+stderr from file dialog helper: [b'', b'']
piped data from file dialog helper: [b'T\xe0\x97&P\xb2\xf5\xf3"2y\x90\x15[4\x89\xe3\xf4\xa6\xce\xfc\xe1O\xd6\xbf9(\x11\xa1z\x84d', b'D:\\Users\\Denis\\DenisDocs\\My Digital Editions\\The Last Dance Behind the Scenes at the Final Four.epub']
DeDRM v7.2.1: Trying to decrypt The Last Dance Behind the Scenes at the Final Four.epub
DeDRM v7.2.1: Verifying zip archive integrity
DeDRM v7.2.1: The Last Dance Behind the Scenes at the Final Four.epub is a secure Adobe Adept ePub
DeDRM v7.2.1: Trying Encryption key default_key
Could not decrypt 6g4cs7cr.epub because of an exception:
Traceback (most recent call last):
  File "calibre_plugins.dedrm.ineptepub", line 418, in decryptBook
TypeError: object of type 'int' has no len()

DeDRM v7.2.1: Failed to decrypt with key default_key after 0.1 seconds
DeDRM v7.2.1: Looking for new default Adobe Digital Editions Keys after 0.1 seconds
DeDRM v7.2.1: Exception when getting default Adobe Key after 0.1 seconds
Traceback (most recent call last):
  File "calibre_plugins.dedrm.__init__", line 358, in ePubDecrypt
  File "calibre_plugins.dedrm.adobekey", line 362, in adeptkeys
  File "calibre_plugins.dedrm.adobekey", line 342, in CryptUnprotectData
calibre_plugins.dedrm.adobekey.ADEPTError: Failed to decrypt user key key (sic)
DeDRM v7.2.1: Ultimately failed to decrypt after 0.1 seconds. Read the FAQs at Harper's repository: https://github.com/apprenticeharper/DeDRM_tools/blob/master/FAQs.md
Added The Last Dance: Behind the Scenes at the Final Four to db in: 4.9
Added 1 books in 5.3 seconds
Running file type plugin DeDRM failed with traceback:
Traceback (most recent call last):
  File "calibre\customize\ui.py", line 176, in _run_filetype_plugins
  File "calibre_plugins.dedrm.__init__", line 644, in run
  File "calibre_plugins.dedrm.__init__", line 415, in ePubDecrypt
calibre_plugins.dedrm.DeDRMError: DeDRM v7.2.1: Ultimately failed to decrypt after 0.1 seconds. Read the FAQs at Harper's repository: https://github.com/apprenticeharper/DeDRM_tools/blob/master/FAQs.md
Exception ignored in: <function NativeFunction.__del__ at 0x00000225837A63A0>
Traceback (most recent call last):
  File "calibre_plugins.dedrm.adobekey", line 268, in __del__
TypeError: 'NoneType' object is not callable
Exception ignored in: <function NativeFunction.__del__ at 0x00000225837A63A0>
Traceback (most recent call last):
  File "calibre_plugins.dedrm.adobekey", line 268, in __del__
TypeError: 'NoneType' object is not callable
Worker Launch took: 0.02 seconds

**_2) DETAILS OF DECRYPT FAILURE GENERATED IN CALIBRE WINDOW "ERROR: Loading book failed._**

calibre, version 5.24.0
ERROR: Loading book failed: Failed to open the book at D:\Users\Denis\DenisDocs\Calibre Library\John Feinstein\The Last Dance_ Behind the Scenes a (729)\The Last Dance_ Behind the Scen - John Feinstein.epub. Click "Show details" for more info.

Failed to convert book: D:\Users\Denis\DenisDocs\Calibre Library\John Feinstein\The Last Dance_ Behind the Scenes a (729)\The Last Dance_ Behind the Scen - John Feinstein.epub with error:
DeDRM v7.2.1: Trying to decrypt The Last Dance_ Behind the Scen - John Feinstein.epub
DeDRM v7.2.1: Verifying zip archive integrity
DeDRM v7.2.1: The Last Dance_ Behind the Scen - John Feinstein.epub is a secure Adobe Adept ePub
DeDRM v7.2.1: Trying Encryption key default_key
Could not decrypt 74za7ms6.epub because of an exception:
Traceback (most recent call last):
  File "calibre_plugins.dedrm.ineptepub", line 418, in decryptBook
TypeError: object of type 'int' has no len()

DeDRM v7.2.1: Failed to decrypt with key default_key after 0.1 seconds
DeDRM v7.2.1: Looking for new default Adobe Digital Editions Keys after 0.1 seconds
DeDRM v7.2.1: Exception when getting default Adobe Key after 0.1 seconds
Traceback (most recent call last):
  File "calibre_plugins.dedrm.__init__", line 358, in ePubDecrypt
  File "calibre_plugins.dedrm.adobekey", line 362, in adeptkeys
  File "calibre_plugins.dedrm.adobekey", line 342, in CryptUnprotectData
calibre_plugins.dedrm.adobekey.ADEPTError: Failed to decrypt user key key (sic)
DeDRM v7.2.1: Ultimately failed to decrypt after 0.1 seconds. Read the FAQs at Harper's repository: https://github.com/apprenticeharper/DeDRM_tools/blob/master/FAQs.md
InputFormatPlugin: EPUB Input running
on D:\Users\Denis\DenisDocs\Calibre Library\John Feinstein\The Last Dance_ Behind the Scenes a (729)\The Last Dance_ Behind the Scen - John Feinstein.epub
Failed to run pipe worker with command: from calibre.srv.render_book import viewer_main; viewer_main()
Running file type plugin DeDRM failed with traceback:
Traceback (most recent call last):
  File "calibre\customize\ui.py", line 176, in _run_filetype_plugins
  File "calibre_plugins.dedrm.__init__", line 644, in run
  File "calibre_plugins.dedrm.__init__", line 415, in ePubDecrypt
calibre_plugins.dedrm.DeDRMError: DeDRM v7.2.1: Ultimately failed to decrypt after 0.1 seconds. Read the FAQs at Harper's repository: https://github.com/apprenticeharper/DeDRM_tools/blob/master/FAQs.md
Traceback (most recent call last):
  File "runpy.py", line 194, in _run_module_as_main
  File "runpy.py", line 87, in _run_code
  File "site.py", line 82, in <module>
  File "site.py", line 77, in main
  File "site.py", line 49, in run_entry_point
  File "calibre\utils\ipc\worker.py", line 197, in main
  File "<string>", line 1, in <module>
  File "calibre\srv\render_book.py", line 824, in viewer_main
  File "calibre\srv\render_book.py", line 815, in render_for_viewer
  File "calibre\srv\render_book.py", line 792, in render
  File "calibre\ebooks\oeb\iterator\book.py", line 63, in extract_book
  File "calibre\customize\conversion.py", line 245, in __call__
  File "calibre\ebooks\conversion\plugins\epub_input.py", line 290, in convert
calibre.ebooks.DRMError: The Last Dance_ Behind the Scen - John Feinstein.epub
Exception ignored in: <function NativeFunction.__del__ at 0x0000022C9D1510D0>
Traceback (most recent call last):
  File "calibre_plugins.dedrm.adobekey", line 268, in __del__
TypeError: 'NoneType' object is not callable
Exception ignored in: <function NativeFunction.__del__ at 0x0000022C9D1510D0>
Traceback (most recent call last):
  File "calibre_plugins.dedrm.adobekey", line 268, in __del__
TypeError: 'NoneType' object is not callable

calibre, version 5.24.0

**Comment by dwall115** [comment]
In setting up a new machine, I've installed the latest Calibre, and 7.2.1 of DeDRM, however no default key exists for Adobe Digital Editions, and I can download and see/read DRM books in Adobe, but for some reason they won't decrypt in Calibre.  This sounded like your issue, but wondered if it's because no decrypt key exists (default or otherwise) for Adobe books in DeDRM?

**Comment by DRWolff1954** [comment]
This does sound the same. I can read the books in ADE. But when added to 
Calibre, the DRM is not removed.
Denis

**Comment by dwall115** [comment]
Precisely.  I’ve noticed that after setting up my new machine, there is no default key in the adobe books for DeDRM, and I have no idea how to create one.  Obviously copying the default key from another machine doesn’t work (I tried it), as each machine will have a unique key based upon that machines id and authorization from Adobe.  If there is a way to create keys using DeDRM, I’m not finding it.  Hopefully we’ll get any answer from the techies at the site.

 

Thanks for confirming.

 

Regards,

d

**Comment by DRWolff1954** [comment]
Which version of ADE are you using?

At one point I tried removing Calibre, ADE and DeDRM from my computer, and 
the reinstalling with older versions. Didn't work. Used to. But now, 
nothing does. Puzzling.
Denis

**Comment by dwall115** [comment]
Currently 4.5 but changed back to 2.01 to try and see if that worked.  It didn’t.  MY wife’s machine is at 4.5 ADE, and latest 5.24 Calibre with 7.21 DeDRM and works great on any and all ADE books.  Like I said earlier, I think somehow the default ADE key is missing when I built this machine and I have no idea how to create one from scratch.  The maddening thing is that when I download an ADE book it opens fine in Adobe and I can read it there.  However I read (normally) 5 or 6 concurrent books, so the 14 day time limit is a bit short.  Believe it or not, I do delete the books once I finish them.  DeDRM simply gives me a bit more time to read them.

 

Thanks for checking in.

Regards,

d

**Comment by DRWolff1954** [comment]
So, on your wife's computer, ADE 4.5 + Calibre 5.24 +  7.21 DeDRM 7.21
works just as expected. But not on yours?

How very very odd. But useful, maybe, to technical support, to compare the
similarities and differences.

I purchase eBooks from several sources (but mostly Kobo). I have been
consolidating them in Calibre to manage my library. Frustrating to now have
my library split into two parts.
Denis

**Comment by dwall115** [comment]
Yep.  You got it.  That is the puzzle I’ve been dealing with for the last 2 weeks.  Hopefully we’ll see some direction at some point.

 

Take care,

d

**Reply by DRWolff1954** [reply]
In the hope that this may help a more technical minded person solve my problem...
As mentioned above, I am currently using ADE 4.5. I have discovered that while ADE 4.5 works fine with new downloads, it will not open older downloads (e.g. added 30/03/2021 or before) that are in the My Digital Editions folder. When I try, I get the following error message:
     Adobe Digital Editions
     An Error has been Detected
     Activation Server error
     code: E_ADEPT_INTERNAL

I suspect that this is because of one of two reasons:
1) I am pretty sure that I was previously using an older version of ADE, probably 2.0.1.
2) I originally had ADE installed without creating an ADE ID. 

It makes me wonder whether the problems that I and others are having originates more on the ADE side than on the DeDRM side. But I have no real idea. Hopefully someone else will.
Denis

**Comment by DRWolff1954** [comment]
I was experimenting and may have found a work-around that works for those who use the Kobo Desktop.

The FAQ had two interesting pieces of advice: 
1) Use an older version of Adobe Digital Editions. So I downgraded from ADE 4.5 to ADE to 2.0.1.78765.
2) Download Kobo ePubs to the Kobo Desktop, and use the Obok plug-in for Calibre to add books and remove DRM. I had never tried this before. Previously, the standard DeDRM worked fine. And actually I wish it still did - I find it easier to work with than Obok. But beggars can't be choosers. 

Anyway, I tried it. I installed Kobo Desktop for Windows, and downloaded the books I had been having trouble with. Using Calibre 5.24 (64 bit), Obok DeDRM 7.2.1, and ADE 2.01, it all seems to work. Clicking on the Obok plug-in automatically identifies all new books, and creates a copy in Calibre. And as far as I can tell, removes DRM at the same time. 

Like I said, a work-around. But maybe it will give the technies a clue about what it doing on.
Denis

**Comment by ElleKayEm** [comment]
Sounds like issues with your Adobe ID authorization.  See instructions here for de-authorizing and re-authorizing ADE (scroll down for Windows):  https://helpx.adobe.com/digital-editions/kb/activation-errors-adobe-digital-editions.html

**Comment by dwall115** [comment]
Although I’ve done that several times, I’ll do it again.

 

In the past, just deauthorizing and reauthorizing didn’t do the trick, so I’ve cleaned the registry of any prior Adobe info before reauthorizing.

 

Thank you,

d

**Comment by dwall115** [comment]
I de-authorized the pc (per instructions) AND removed the entry in the registry (per instructions), re-authorized the pc, and still no default key in the Adobe customization panel in the plug in.

 

Regards,

d

**Comment by ElleKayEm** [comment]
After re-authorizing, did you download a fresh copy of the book?  Delete the old book from calibre and import the fresh copy?  A key is saved in DeDRM if a book is successfully decrypted with it.

**Comment by dwall115** [comment]
Not yet.  I’ll give that a try shortly.  Thank you.

d

**Comment by DRWolff1954** [comment]
Just wondering whether it worked?
Thanks
Denis

**Comment by dwall115** [comment]
I didn’t realize, but my library card expired, hence the ability to download ebooks until it’s renewed.  I completed the renewal info on Friday, but am not anticipating it will be renewed until tomorrow.  At which time I’ll snag a test book, and let everyone know if indeed that worked.  The only thing differently I did this time (unauthorizing and reauthorizing the machine) was to delete the entry in the registry for adept registration.  I was unaware that the thing that put the default key in place was the first book de-drm’d with the tool.

 

Again, sorry for the delay in the response, but I will update once I can again download e-books.  Hopefully by Tuesday this week.

 

Regards,

d

**Comment by dwall115** [comment]
Hey everyone!  Great news!

 

I just grabbed and Adobe book, imported into Calibre and it de-drm’d just fine!  Let me be clear here;

I’m using Adobe Digital Editions 4.5, Calibre 5.27 and DeDrm plug-in 7.21.  The ONLY thing I did differently this time was to REMOVE the registry setting per the instructions with Regedit before re-registering my copy of Digital Editions to this machine.  This is VERY good news for those of us that were banging our heads against the wall trying to figure out what was wrong.  And to point out the fact I just learned via this entire process, once a book is successfully imported into Calibre, that is what creates the default key in the plugin, not the other way around.

 

Thanks to all for the assistance, and good luck to all the others out there attempting to get the plugin working again after moving to a new machine.

 

Best regards,

d

 

From: ***@***.*** ***@***.***> 
Sent: Sunday, September 12, 2021 6:25 AM
To: 'apprenticeharper/DeDRM_tools' ***@***.***>; 'apprenticeharper/DeDRM_tools' ***@***.***>
Cc: 'Manual' ***@***.***>; 'Doug Wallace' ***@***.***>
Subject: RE: [apprenticeharper/DeDRM_tools] DeDRM Decryption of EPUB in Calibre failed - Exception when getting Adobe Key (Discussion #1798)

 

I didn’t realize, but my library card expired, hence the ability to download ebooks until it’s renewed.  I completed the renewal info on Friday, but am not anticipating it will be renewed until tomorrow.  At which time I’ll snag a test book, and let everyone know if indeed that worked.  The only thing differently I did this time (unauthorizing and reauthorizing the machine) was to delete the entry in the registry for adept registration.  I was unaware that the thing that put the default key in place was the first book de-drm’d with the tool.

 

Again, sorry for the delay in the response, but I will update once I can again download e-books.  Hopefully by Tuesday this week.

 

Regards,

d
